shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1596,7 +1597,56 @@ - (void)startKeyBoardAnimation:(NSTimeInterval)duration {
 
   // Invalidate old vsync client if old animation is not completed.
   [self invalidateKeyboardAnimationVSyncClient];
-  [self setupKeyboardAnimationVsyncClient];
+
+  fml::WeakPtr<FlutterViewController> weakSelf = [self getWeakPtr];
+  FlutterKeyboardAnimationCallback keyboardAnimationCallback = ^(
+      fml::TimePoint keyboardAnimationTargetTime) {
+    if (!weakSelf) {
+      return;
+    }
+    fml::scoped_nsobject<FlutterViewController> flutterViewController(
+        [(FlutterViewController*)weakSelf.get() retain]);
+    if (!flutterViewController) {
+      return;
+    }
+
+    // If the view controller's view is not loaded, bail out.
+    if (flutterViewController.get().viewIfLoaded == nil) {
+      return;
+    }
+    // If the view for tracking keyboard animation is nil, means it is not
+    // created, bail out.
+    if ([flutterViewController keyboardAnimationView] == nil) {
```

Comment:
not familiar with the keyboard logic, so maybe dumb question - what's the purpose of keyboardAnimationView? what kind of animation?

Reply:
`keyboardAnimationView` is a view to let UIKit execute animation. If we don't add this in view hierarchy, then the UIVIew animation block will be invaild because UIKit think there is no necessary to add animation, so we can't tracking animation completion time.